posix/fdtable.c

```diff
@@ -0,0 +1,232 @@
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > OE_INT_MAX)
+        new_size = OE_INT_MAX;
+
+    if (new_size > _table_size)
+    {
+        entry_t* p;
+        size_t n = new_size;
+
+        /* Reallocate the table. */
+        if (!(p = oe_realloc(_table, n * sizeof(entry_t))))
+            goto done;
+
+        /* Zero-fill the unused porition. */
+        {
+            const size_t num_bytes = (n - _table_size) * sizeof(entry_t);
+
+            if (oe_memset_s(p + _table_size, num_bytes, 0, num_bytes) != OE_OK)
+                goto done;
+        }
+
+        _table = p;
+        _table_size = new_size;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+/*
+**==============================================================================
+**
+** Public interface:
+**
+**==============================================================================
+*/
+
+int oe_fdtable_assign(oe_device_t* device)
+{
+    int ret = -1;
+    size_t index;
+
+    oe_spin_lock(&_lock);
+
+    if (!device)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* The file-descriptor table must be big enough for standard devices. */
+    if (_table_size < OE_STDERR_FILENO + 1)
+    {
+        if (_resize_table(OE_STDERR_FILENO + 1) != 0)
+        {
+            oe_errno = OE_ENOMEM;
+            goto done;
+        }
+    }
+
+    /* Search for a free slot in the file descriptor table. */
+    for (index = OE_STDERR_FILENO + 1; index < _table_size; index++)
+    {
+        if (!_table[index])
+            break;
+    }
+
+    /* If no free slot found, expand size of the file descriptor table. */
+    if (index == _table_size)
+    {
+        if (_resize_table(_table_size + 1) != 0)
+            OE_RAISE_ERRNO(OE_ENOMEM);
+    }
+
+    _table[index] = device;
+    ret = (int)index;
+
+done:
+
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+int oe_fdtable_release(int fd)
+{
+    int ret = -1;
+
+    oe_spin_lock(&_lock);
+
+    /* Fail if fd is out of range. */
+    if (!(fd >= 0 && (size_t)fd < _table_size))
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    /* Fail if entry was never assigned. */
+    if (!_table[fd])
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    _table[fd] = NULL;
+
+    ret = 0;
+
+done:
+
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+int oe_fdtable_reassign(int fd, oe_device_t* device)
+{
+    int ret = -1;
+
+    oe_spin_lock(&_lock);
+
+    _resize_table(TABLE_CHUNK_SIZE);
+
+    if (fd < 0 || (size_t)fd >= _table_size)
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    if (_table[fd])
+    {
+        if (OE_CALL_BASE(close, _table[fd]) != 0)
+            OE_RAISE_ERRNO(oe_errno);
+    }
+
+    /* Set the device. */
+    _table[fd] = device;
+
+    ret = 0;
+
+done:
+
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+static oe_device_t* _get_fd_device(int fd)
+{
+    oe_device_t* ret = NULL;
+
+    oe_spin_lock(&_lock);
+
+    if (fd < 0 || fd >= (int)_table_size)
+        OE_RAISE_ERRNO(OE_EBADF);
```

Comment:
Upcast `fd` instead of downcasting `_table_size`, not sure why you chose to do that here unlike the other methods.

For my understanding, what if someone calls `oe_fdtable_get` for a standard interface without an assign call? Where is the default table initialized?

Reply:
Thanks. I missed that one during my review.